lib/customisation.nix

```
@@ -404,4 +405,96 @@ rec {
       };
     in self;
 
+  /*
+    Define a `mkDerivation`-like function based on another `mkDerivation`-like function.
+
+    `mkDerivation` gives access to its final set of derivation attributes when
+    it is passed a function, or when it is passed an overlay-style function in
+    `overrideAttrs`.
+
+    Instead of composing new `mkDerivation`-like build helpers using normal
+    function composition, `extendMkDerivation` makes sure that the returned
+    build helper supports such first class recursion like `mkDerivation` does.
+
+    the base build helper by changing the input arguments. The choice of the
+    base build helper must be `finalAttrs`-independent. In most cases,
+    *getMkDerivationBase* is simply a constant function that returns the base
+    build helper.
+
+    This functions are mainly about defining build helpers (`a == Derivation`),
+    but is flexible enough to handle return types other than derivation.
+
+    Type:
+      extendMkDerivation ::
+        (FixedPointArgs -> (FixedPointArgs | AttrSet) -> a)
+        -> (AttrSet -> AttrSet -> AttrSet)
+        -> (FixedPointArgs | AttrSet) -> a
+
+      FixedPointArgs :: AttrSet -> AttrSet
+
+    Example:
+      mkLocalDerivation = lib.extendMkDerivation pkgs.stdenv.mkDerivation (finalAttrs:
+        args@{ preferLocalBuild ? true, allowSubstitute ? false, impassablePredicate ? (_: false), ... }:
+        removeAttrs args [ "impassablePredicate" ] // { inherit preferLocalBuild allowSubstitute; })
+
+      mkLocalDerivation.__functionArgs
+      => { allowSubstitute = true; impassablePredicate = true; preferLocalBuild = true; }
+
+      mkLocalDerivation { inherit (pkgs.hello) pname version src; impassablePredicate = _: false; }
+      => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+      mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; impassablePredicate = _: false; })
+      => «derivation /nix/store/xirl67m60ahg6jmzicx43a81g635g8z8-hello-2.12.1.drv»
+
+      (mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; passthru = { foo = "a"; bar = "${finalAttrs.passthru.foo}b"; } })).bar
+      => "ab"
+  */
+  extendMkDerivation =
+    # mkDerivation-like build helper to extend
+    mkDerivationBase:
+    # Overlay implementation of the new build helper
+    impl:
+    mirrorFunctionArgs
+      # Make the __functionArgs looks like one from a build helper accepting plain attribute set.
+      (impl { })
+      # Adds the fixed-point style support
+      (fpargs: mkDerivationBase (finalAttrs:
+        impl finalAttrs (if isFunction fpargs then fpargs finalAttrs else fpargs)
+      ));
+
+
+  /*
+    Like `extendMkDerivation`, but additionally accepts a function
+    to be applied to the result derivation.
+
+    :::{.note}
+    The derivation modification function *modify* should take care of
+    existing attributes that performs overriding (e.g. `<pkg>.overrideAttrs`),
```

```suggestion
    existing attributes that perform overriding (e.g. `<pkg>.overrideAttrs`),
```

I'm not sure what you want to say here though. What does "take care" and "perform overriding" mean here?

---

Sorry for my literacy of documentation writing.

`<pkg>.override`, `<pkg>.overrideAttrs`, `<pkg>.overrideDerivation` and `<pkg>.overridePythonAttrs` are examples of "attributes that performs overriding".

To add an attribute that performs overriding to a derivation that already has such kind of attributes, each of the existing ones need to be updated to re-apply the changes to play well with the new attribute. The following shows how it is done in `lib.makeOverridable`,

```nix
{ lib }:
{
  makeOverridable = f: origArgs:
    let
      # ...
      result = ...;
      overrideResult = g: makeOverridable (args: g (f args)) origArgs;
  in
  if builtins.isAttrs result then
    result // lib.optionalAttrs (result?overrideAttrs) {
      overrideAttrs = fdrv: overrideResult (r: (r.overrideAttrs fdrv));
   }
  else ...;
}
```

The `overrideResult` abstraction is introduced in e140d709c43ab25d993c5c78b4e01aee754b49af in #67809. See https://github.com/NixOS/nixpkgs/blob/a4896cb4aab132c0bcc32c5df4440f6d3b7ec5de/lib/customisation.nix#L85-L87 for detailed implementation.

#267296 makes `overridePythonAttrs` work together with `overrideAttrs` using similar tricks.

One should change a derivation by overriding as possible, any other ways of attribute updating are deemed to be lost after overriding, unless every existing attributes that perform overriding are taken care of like above. This is why `lib.extendDerivation` is not suitable for packages returned by `stdenv.mkDerivation`. #269785 introduces an alternative that takes care of these attributes.